### ISSUE TITLE
Removing a Display causes a VT switch

### ISSUE THREAD

**Issue by arrowd** [sddm/sddm#1803]
I bumped into a following problem:

1. SDDM starts and creates the first Display (vt9 in my case).
2. Greeter runs on this Display.
3. User logs in and the session replaces the greeter. The Display stays the same.
4. User performs "Switch user" action.
5. New Display is created (vt10) and Greeter is shown there.
6. Log into different user. Like in a step 3, Greeter exits and a user session occupies the Display.
7. We now have 2 Displays both running a user session.
8. User performs "Switch user" action again.
9. The third Display is created and Greeter runs in it.
10. Login into the first user again.

Step 10 triggers the "reuse" code path and causes SDDM to switch into vt9. After a small delay SDDM shuts down itself. The `removeDisplay(display)` call in `Seat::displayStopped()` causes the third Display to be destroyed and the third `Xorg` process to exit. When it exits the VT switch happens which always brings me to the **last occupied VT**. Effectively, after a few seconds of being in the user session on vt9 I get switched to the other session (vt10) which is in a locked state.

That Xorg shutdown behavior might be FreeBSD specific, so I'm not sure if it is visible on Linux.

**Comment by Vogtinator** [comment]
IIRC I hit this as well while trying to improve session switching. There are a LOT of race conditions.

Back then I tried to start X with `-novtswitch` to make this more explicit and controllable, but it only ended up breaking stuff even more.

**Comment by arrowd** [comment]
Do you have an idea how to move this forward? I want to fix this, so any input would be appreciated.

> Back then I tried to start X with -novtswitch to make this more explicit and controllable, but it only ended up breaking stuff even more.

How exactly did it break? Does it also disable VT switch on Xorg start?

**Comment by arrowd** [comment]
What if we don't remove a Display if we switch into an existing session? This is suboptimal, as it would cause SDDM greeter to work in the background, but at least would get us rid of switching problems.

**Comment by Vogtinator** [comment]
> Do you have an idea how to move this forward? I want to fix this, so any input would be appreciated.
> 
> > Back then I tried to start X with -novtswitch to make this more explicit and controllable, but it only ended up breaking stuff even more.
> 
> How exactly did it break? Does it also disable VT switch on Xorg start?

I don't think so, but this actually shouldn't be much of an issue as SDDM handles VT switching already.

> What if we don't remove a Display if we switch into an existing session? This is suboptimal, as it would cause SDDM greeter to work in the background, but at least would get us rid of switching problems.

IMO the best option to fix this is to handle display stopping better. With #1692 this should be mostly (?) done, but there are edge cases.

Could you just try `-novtswitch` combined with #1692?

Meanwhile I found https://github.com/sddm/sddm/issues/1281, which is basically just this issue.

**Comment by arrowd** [comment]
Ok, so first I had to actually implement `-novtswitch` support in the Xorg itself for FreeBSD.

Then I tried it without #1692 and it still wasn't working right because the code in `displayStopped` was sometimes switching to the wrong VT.

Now with both `-novtswitch` and #1692 it finally seems to work corrrectly. I successfully switch back and forth between 2 sessions multiple times.

The choice to switch back to the another session when the current one exits is a bit questional, though. I was expecting to be put into SDDM greeter again, but maybe it is just me. Anyways, the change in #1692 looks very useful.

**Comment by Vogtinator** [comment]
Awesome!

> Ok, so first I had to actually implement -novtswitch support in the Xorg itself for FreeBSD.

Did the option do nothing or did X error out if it was not present? If it's an error we might have to do runtime detection or make it configurable but disabled by default.

> The choice to switch back to the another session when the current one exits is a bit questional, though. I was expecting to be put into SDDM greeter again, but maybe it is just me.

I agree with that design: Starting a greeter would only be useful if a session for a third user is meant to be opened after logging out the second, which isn't that likely IMO. Optimizing for the likely case and not starting the greeter saves the user time and the system memory and CPU.

**Comment by arrowd** [comment]
> Did the option do nothing or did X error out if it was not present?

The former. Its handling is located in the OS-specific parts of the tree (`hw/xfree86/os-support/`) so I took a look at the Linux bits and repeated the idea for FreeBSD.

The important thing about `-novtswitch` is that when SDDM spawns its first screen, we should avoid passing this flag. Otherwise, shutting down the SDDM service (`service sddm stop` or whatever it is for systemd) would leave the screen at the empty VT.

> Optimizing for the likely case and not starting the greeter saves the user time and the system memory and CPU.

Yes, this also makes sense. Anyways, it is pretty easy to change in the code.

**Comment by Vogtinator** [comment]
> > Did the option do nothing or did X error out if it was not present?
> 
> The former. Its handling is located in the OS-specific parts of the tree (`hw/xfree86/os-support/`) so I took a look at the Linux bits and repeated the idea for FreeBSD.

Ok, so we could just use it and assume it works. Unless something weird happens, the worst is that the same VT switch happens twice.

> The important thing about `-novtswitch` is that when SDDM spawns its first screen, we should avoid passing this flag. Otherwise, shutting down the SDDM service (`service sddm stop` or whatever it is for systemd) would leave the screen at the empty VT.

That needs to be handled differently and explicitly in SDDM shutdown code for two reasons:
1. SDDM can be smarter and query the system for other active sessions. X would just switch to the (effectively random) old one
2. The first X session is not necessarily the last one to quit. If you start a second one and log out from the first, only the second X server without `-novtswitch` remains.

**Comment by grahamperrin** [comment]
<https://github.com/sddm/sddm/issues/1803#issue-1915992141>

> … 10. Login into the first user again.

If I'm not mistaken, this is no longer possible: 

- after the second user logs in, the switch dialogue presents the second user alone
- the first user can not regain access (to the original session) until the second user logs out. 

----

If the screen locks during the second session, and if the second user is absent when the first user aims to regain access, then the first user is effectively locked out by the absentee. 

### Postscript

Thanks to @arrowd for direction in IRC: 

- FreeBSD bug [241231 – security/plasma5-kscreenlocker: "Start New Session" does not work](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=241231)

**Comment by arrowd** [comment]
Your context (FreeBSD) is not relevant to this issue.

**Comment by Vogtinator** [comment]
> Now with both -novtswitch and https://github.com/sddm/sddm/pull/1692 it finally seems to work corrrectly. I successfully switch back and forth between 2 sessions multiple times.

I opened a PR that includes `-novtswitch`: #1968
